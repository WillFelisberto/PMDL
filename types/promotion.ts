export type ConditionType =
  | 'comparator'
  | 'iterator'
  | 'group-iterator'
  | 'quantifier'
  | 'operator'
  | 'operation';

export type ConditionBase = {
  id: string;
  type: ConditionType;
  attributes?: Array<{ name: string; value: string }>;
};

export type ComparatorCondition = ConditionBase & {
  type: 'comparator';
  comparatorType:
    | 'equals'
    | 'not-equals'
    | 'less-than'
    | 'less-than-or-equals'
    | 'greater-than'
    | 'greater-than-or-equals'
    | 'contains'
    | 'starts-with'
    | 'ends-with'
    | 'isoneof'
    | 'isnotoneof'
    | 'includes'
    | 'includes-any'
    | 'includes-all';
  leftValue: string;
  rightValue: string;
  dataType: string;
};

export type ValueType = {
  type: 'value' | 'constant' | 'array' | 'null';
  dataType?: string;
  value?: string | number | boolean;
  values?: ValueType[];
};
export type IteratorCondition = ConditionBase & {
  type: 'iterator';
  iteratorType: 'next' | 'every' | 'up-to-and-including' | 'custom';
  collectionName: string;
  elementName: string;
  number?: number;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  condition?: PromotionCondition;
};

export type GroupIteratorCondition = ConditionBase & {
  type: 'group-iterator';
  iteratorType: 'every-group' | 'up-to-and-including-group' | 'custom';
  number?: number;
  conditions?: PromotionCondition[];
};

export type QuantifierCondition = ConditionBase & {
  type: 'quantifier';
  quantifierType: 'exactly' | 'at-least' | 'at-most' | 'all' | 'custom';
  number?: number;
  collectionName: string;
  elementName: string;
  condition?: PromotionCondition;
  customType?: string;
};

export type OperatorCondition = ConditionBase & {
  type: 'operator';
  operatorType: 'and' | 'or' | 'not' | 'custom';
  conditions: PromotionCondition[];
  customType?: string; // Para operadores personalizados
};

export type OperationCondition = ConditionBase & {
  type: 'operation';
  operationType: 'union' | 'anded-union' | 'custom';
  iterators: PromotionCondition[];
  customType?: string;
};
export type PromotionCondition =
  | ComparatorCondition
  | IteratorCondition
  | GroupIteratorCondition
  | QuantifierCondition
  | OperatorCondition
  | OperationCondition;

export type DiscountStructure = {
  calculatorType: string;
  discountType: 'percentOff' | 'amountOff' | 'fixedPrice';
  adjuster: number;
  target?: 'order' | 'shipping' | 'item';
};

export type Offer = {
  discountStructures: DiscountStructure[];
};

export type Promotion = {
  displayName: string;
  description: string;
  enabled: boolean;
  priority: number;
  qualifier: PromotionCondition | undefined;
  offer: Offer;
  conditions: PromotionCondition[];
  sites: Array<{ repositoryId: string }>;
};
