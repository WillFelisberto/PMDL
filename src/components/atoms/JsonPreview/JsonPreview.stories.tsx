import type { Meta, StoryObj } from '@storybook/react';
import { JsonPreview } from './JsonPreview';
import { Promotion } from '@/types/promotion';

const meta: Meta<typeof JsonPreview> = {
  title: 'Atoms/JsonPreview',
  component: JsonPreview,
  tags: ['autodocs'],
  argTypes: {
    promotion: {
      control: 'object',
      description: 'Objeto de promoção usado para gerar o JSON'
    }
  }
};

export default meta;

type Story = StoryObj<typeof JsonPreview>;

const samplePromotion: Promotion = {
  displayName: 'Promoção de Teste',
  description: 'Descrição da promoção de teste',
  priority: 1,
  offer: {
    discountStructures: [
      {
        target: 'order',
        calculatorType: 'default',
        discountType: 'percentOff',
        adjuster: 10
      }
    ]
  },
  enabled: false,
  qualifier: undefined,
  conditions: [],
  sites: [
    {
      repositoryId: 'site-1'
    }
  ]
};

export const Default: Story = {
  args: {
    promotion: samplePromotion
  }
};
