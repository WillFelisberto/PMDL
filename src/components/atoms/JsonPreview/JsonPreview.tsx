'use client';
import { Promotion } from '@/types/promotion';
import { Prism as ReactSyntaxHighlighter } from 'react-syntax-highlighter';
import { atomDark } from 'react-syntax-highlighter/dist/cjs/styles/prism';
import { formatXmlToSingleLine } from 'utils/formatXML';
import { generatePmdl } from 'utils/generatePmdl';

type JsonPreviewProps = {
  promotion: Promotion;
};

export const JsonPreview = ({ promotion }: JsonPreviewProps) => {
  const isValidPromotion = () => {
    return (
      promotion.displayName.trim() !== '' &&
      promotion.description.trim() !== '' &&
      promotion.priority !== null &&
      promotion.offer.discountStructures.length > 0 &&
      promotion.offer?.discountStructures?.[0]?.target?.trim() !== ''
    );
  };

  const generateOccPayload = () => {
    return {
      displayName: promotion.displayName,
      description: promotion.description,
      enabled: true,
      priority: promotion.priority,
      templatePath: promotion.offer.discountStructures[0].target,
      templateName: 'rawPmdlTemplate',
      sites: [
        {
          repositoryId: 'siteUS'
        }
      ],
      templateValues: {
        pmdl: {
          xml: formatXmlToSingleLine(generatePmdl(promotion))
        }
      }
    };
  };

  // SÃ³ exibe o JSON se todos os campos estiverem preenchidos
  if (!isValidPromotion()) {
    return null;
  }

  const jsonString = JSON.stringify(generateOccPayload(), null, 2);

  return (
    <div className="mt-8 space-y-4">
      <h3 className="text-lg font-semibold text-gray-700">Payload OCC (JSON)</h3>

      <div className="rounded-lg overflow-hidden border border-gray-200 relative">
        <button
          data-testid="copy-button"
          onClick={() => navigator.clipboard.writeText(jsonString)}
          className="absolute right-2 top-2 bg-gray-800 text-white px-3 py-1 rounded-md text-sm hover:bg-gray-700 transition-colors"
        >
          Copiar
        </button>

        <ReactSyntaxHighlighter
          language="json"
          style={atomDark}
          data-testid="json-preview"
          customStyle={{
            padding: '1.5rem',
            fontSize: '0.875rem',
            lineHeight: '1.5',
            margin: 0
          }}
        >
          {jsonString}
        </ReactSyntaxHighlighter>
      </div>
    </div>
  );
};
