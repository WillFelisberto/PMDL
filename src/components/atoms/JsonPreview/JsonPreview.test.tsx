import { JsonPreview } from '@/components/atoms/JsonPreview';
import { render, screen, fireEvent } from '@/tests/test-utils';
import { Promotion } from '@/types/promotion';

jest.mock('@/utils/generatePmdl', () => ({
  generatePmdl: jest.fn(() => '<pmdl>Mocked XML</pmdl>')
}));

describe('JsonPreview Component', () => {
  // Mock completo e válido
  const validMockPromotion: Promotion = {
    displayName: 'Promoção Teste',
    description: 'Descrição válida',
    enabled: true,
    priority: 1,
    conditions: [],
    qualifier: undefined,
    offer: {
      discountStructures: [
        {
          calculatorType: 'standard',
          discountType: 'percentOff',
          adjuster: 10,
          target: 'order'
        }
      ]
    },
    sites: []
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Deve renderizar corretamente o título do JSON Preview', () => {
    const { container } = render(<JsonPreview promotion={validMockPromotion} />);

    expect(screen.getByText('Payload OCC (JSON)')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('Deve gerar corretamente o JSON com os valores da promoção', () => {
    render(<JsonPreview promotion={validMockPromotion} />);

    const jsonElement = screen.getByTestId('json-preview');

    expect(jsonElement).toHaveTextContent('"displayName": "Promoção Teste"');
    expect(jsonElement).toHaveTextContent('"priority": 1');
    expect(jsonElement).toHaveTextContent('"templatePath": "order"');
    expect(jsonElement).toHaveTextContent('"templateName": "rawPmdlTemplate"');
  });

  it('Deve copiar o JSON ao clicar no botão', async () => {
    Object.assign(navigator, {
      clipboard: {
        writeText: jest.fn()
      }
    });

    render(<JsonPreview promotion={validMockPromotion} />);

    const copyButton = await screen.findByTestId('copy-button');
    fireEvent.click(copyButton);

    expect(navigator.clipboard.writeText).toHaveBeenCalledTimes(1);
    expect(navigator.clipboard.writeText).toHaveBeenCalledWith(
      expect.stringContaining('"displayName": "Promoção Teste"')
    );
  });
});
