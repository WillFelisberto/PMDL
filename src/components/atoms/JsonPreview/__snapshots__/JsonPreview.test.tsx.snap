// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonPreview Component Deve renderizar corretamente o título do JSON Preview 1`] = `
<div>
  <div
    class="mt-8 space-y-4"
  >
    <h3
      class="text-lg font-semibold text-gray-700"
    >
      Payload OCC (JSON)
    </h3>
    <div
      class="rounded-lg overflow-hidden border border-gray-200 relative"
    >
      <button
        class="absolute right-2 top-2 bg-gray-800 text-white px-3 py-1 rounded-md text-sm hover:bg-gray-700 transition-colors"
        data-testid="copy-button"
      >
        Copiar
      </button>
      <pre
        data-testid="json-preview"
        style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none; padding: 1.5rem; margin: 0px; overflow: auto; border-radius: 0.3em; background: rgb(29, 31, 33); font-size: 0.875rem;"
      >
        <code
          class="language-json"
          style="white-space: pre; color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none;"
        >
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            {
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "displayName"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(168, 255, 96);"
          >
            "Promoção Teste"
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ,
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "description"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(168, 255, 96);"
          >
            "Descrição válida"
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ,
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "enabled"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(153, 204, 153);"
          >
            true
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ,
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "priority"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(255, 115, 253);"
          >
            1
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ,
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "templatePath"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(168, 255, 96);"
          >
            "order"
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ,
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "templateName"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(168, 255, 96);"
          >
            "rawPmdlTemplate"
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ,
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "sites"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            [
          </span>
          <span>
            

          </span>
          <span>
                
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            {
          </span>
          <span>
            

          </span>
          <span>
                  
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "repositoryId"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(168, 255, 96);"
          >
            "repository-1"
          </span>
          <span>
            

          </span>
          <span>
                
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            }
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ]
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ,
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "templateValues"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            {
          </span>
          <span>
            

          </span>
          <span>
                
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "pmdl"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            {
          </span>
          <span>
            

          </span>
          <span>
                  
          </span>
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            "xml"
          </span>
          <span
            class="token"
            style="color: rgb(237, 237, 237);"
          >
            :
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(168, 255, 96);"
          >
            "&lt;pmdl&gt;Mocked XML&lt;/pmdl&gt;"
          </span>
          <span>
            

          </span>
          <span>
                
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            }
          </span>
          <span>
            

          </span>
          <span>
              
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            }
          </span>
          <span>
            

          </span>
          <span />
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            }
          </span>
        </code>
      </pre>
    </div>
  </div>
</div>
`;
